### PR TITLE
Improved Payment Details Handling for Partial Captures

### LINKED ISSUE #6721: Fees show they're applied to the Authorize amount instead of Capture amount
### Describe the bug
When the authorize and capture payments feature is enabled, merchants see that the fee is based off the authorized amount. The fee is correctly calculated based on the captured amount, but this is not reflected in the order notes. 

### To Reproduce

1. Go to WooCommerce Payments settings and the Issue an authorization on checkout, and capture later setting.
2. Complete a transaction that authorizes the payment. 

### Actual behavior

- In the order notes and when inspecting the payment ID, the fee shows that it is based off the authorized amount. 

- The fee is correctly calculated based on the capture amount. 

### Screenshots

- The initial authorization is for $108.21, but the capture is for $50. 
- The fee shown ($1.75) is for the captured amount of $50, not the authorized amount, which would have a fee of $3.44 like - in the second photo. 
- The third photo shows that the transaction fee under the order details is correctly reflected both when the payment is authorized and changes when the payment is captured. 

Markup on 2023-07-10 at 10_06_03
Markup 2023-07-10 at 09 58 55
Annotation on 2023-07-10 at 19-25-08

### Expected behavior

- The fee should reflect the correct amount it's applied to. 
- The initial authorized amount should show the fee that is associated with that amount. 
- When the payment is captured later for a different amount, a new note should reflect the new fee, and final fee, that is applied to the captured amount.

### Desktop (please complete the following information):

* OS: 13.4.1 
* Browser Chrome
* Version: WooCommerce Payments v 6.1.1

### PR DESCRIPTION
**Pull request by mgascam**
Fixes #6721 
_Depends on this change in the server_ https://github.com/Automattic/woocommerce-payments-server/pull/4620
#### Changes proposed in this Pull Request
This pull request introduces improvements in handling payment details, focusing on partial captures and foreign exchange (FX) calculations:

1. Client-side Event Mapping (map-events.js): Ensure that displayed amounts reflect the actual captured values, especially in partial capture situations.

2. Captured Event Note Class (class-wc-payments-captured-event-note.php): Same as the front end, ensure that displayed amounts are the ones actually captured in the order notes.

#### Testing instructions

* Coming soon

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `8107` or branch name `fix/6721-fee-applied-to-authorized-amount` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=fix/6721-fee-applied-to-authorized-amount) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: b635718f123f5776b0fe5369ae328391e28c6705
- Build time: 2024-02-13 12:05:58 UTC

_Note: the build is updated when a new commit is pushed to this PR._

**Review by kalessil — CHANGES_REQUESTED** [review]
I didn't dive very deep, but this approach is too complex (both reading and code complexity).

A big chunk of complexity is caused by `amountCaptured ? amountCaptured : ...`,
so we should look into the data structures that the server sends. 

The server could always populate the amount captured: captured amount details can fall back to the amount details and always be present. As the captured amount details are newly introduced on the server side, I'd expect no side effects with this approach, but that needs a deeper dive.

**Comment by mgascam** [comment]
> I didn't dive very deep, but this approach is too complex (both reading and code complexity).
> 
> A big chunk of complexity is caused by `amountCaptured ? amountCaptured : ...`, so we should look into the data structures that the server sends.
> 
> The server could always populate the amount captured: captured amount details can fall back to the amount details and always be present. As the captured amount details are newly introduced on the server side, I'd expect no side effects with this approach, but that needs a deeper dive.

Thanks@kalessil. I'm following your suggestion to use the amount captured, since it will be always defined in server response. The most relevant commits are https://github.com/Automattic/woocommerce-payments/pull/8107/commits/6d77dd1b21a61b5349c6a9e226d210c958c79743 and https://github.com/Automattic/woocommerce-payments/pull/8107/commits/6d77dd1b21a61b5349c6a9e226d210c958c79743 (rest of commits relate to tests).

**Comment by kalessil** [comment]
@mgascam: it works like intended, I noticed only one issue on the order page: on partial capture, the fee meta-data is not getting updated (should be $0.72 in my case):

Screenshot from 2024-02-02 09-53-14

File: includes/class-wc-payments-captured-event-note.php [hunk 1]
```diff
@@ -173,15 +173,33 @@ public function compose_fee_break_down() {
 	public function compose_net_string(): string {
 		$data = $this->captured_event['transaction_details'];
 
-		$net = WC_Payments_Utils::interpret_stripe_amount( (int) $data['store_amount'] - $data['store_fee'], $data['store_currency'] );
+		// Determine the type of payment and select the appropriate amounts and currencies.
+		if ( $this->is_fx_event() ) {
+			// For fx events, we need the store amount and currency to display the net amount
+			// in the store currency.
+			$amount_key          = 'store_amount';
```
Comment (kalessil) [inline includes/class-wc-payments-captured-event-note.php]: It is the last nudge about complexity (and how friendly we are with Static code analysis). We'd better refind the approach in this if-else statement:

```php
    $amount_key = 'store_amount'; // before

    $amount = $data[ 'store_amount' ]; // after
```

Comment (mgascam) [reply includes/class-wc-payments-captured-event-note.php]: Good point, followed your suggestion here https://github.com/Automattic/woocommerce-payments/pull/8107/commits/b7433f38d7648142a233a20a32c75c731a69dca0 thanks!

**Comment by mgascam** [comment]
> @mgascam: it works like intended, I noticed only one issue on the order page: on partial capture, the fee meta-data is not getting updated (should be $0.72 in my case):
> 
> Screenshot from 2024-02-02 09-53-14

Indeed, I found that it gets eventually updated when the async job to add the fee note kicks in. But seems there's also an opportunity to update the fee when the charge is captured since we have the updated charge information, including the fee, so there's no point in making the merchant wait. I'm trying it in this commit https://github.com/Automattic/woocommerce-payments/pull/8107/commits/4e89765941041afc7b3685115cc3677554ccd426. The change itself is quite simple since we only need to call `attach_transaction_fee_to_order` that already takes care of updating the meta with the updated fee amount. You'll notice more changes, but those are just a small refactor needed to provide the `payment_intent` instance to the mentioned method.

File: includes/class-wc-payments-captured-event-note.php [hunk 2]
```diff
@@ -103,7 +103,7 @@ public function compose_fee_string(): string {
 		$fee_rates      = $data['fee_rates'];
 		$percentage     = $fee_rates['percentage'];
 		$fixed_currency = $fee_rates['fixed_currency'];
-		$fixed          = WC_Payments_Utils::interpret_stripe_amount( (int) $fee_rates['fixed'], $fixed_currency );
+		$fixed          = WC_Payments_Utils::interpret_stripe_amount( (int) round( $fee_rates['fixed'] ), $fixed_currency );
```
Comment (kalessil) [inline includes/class-wc-payments-captured-event-note.php]: The fixed part of fees is stored as an integer type, so I'm not sure why we need rounding here. Type casting is a reasonable safeguard, but I'm curious about the rounding.

File: includes/class-wc-payments-captured-event-note.php [hunk 3]
```diff
@@ -173,15 +173,33 @@ public function compose_fee_break_down() {
...
+			$currency        = $data['store_currency'];
+		} else {
+			$amount          = $data['customer_amount'];
+			$captured_amount = $data['customer_amount_captured'];
+			$fee             = $data['customer_fee'];
+			$currency        = $data['customer_currency'];
+		}
+
+		$gross_amount = $captured_amount ?? $amount;
+		$net          = WC_Payments_Utils::interpret_stripe_amount( (int) $gross_amount - $fee, $currency );
```
Comment (kalessil) [inline includes/class-wc-payments-captured-event-note.php]: Suggession: `(int) ($gross_amount - $fee)`

File: client/payment-details/timeline/map-events.js [hunk 4]
```diff
@@ -253,10 +262,12 @@ export const composeFeeString = ( event ) => {
 
 	const {
 		percentage,
-		fixed,
 		fixed_currency: fixedCurrency,
 		history,
 	} = event.fee_rates;
+	// The fixed part of the fee is a float in cents, so we need to prevent rounding issues.
+	const fixed = Math.round( event.fee_rates.fixed );
```
Comment (kalessil) [inline client/payment-details/timeline/map-events.js]: Also, the fixed part is supposed to be an integer. Do we know the trail where floats are getting in?

File: includes/class-wc-payments-order-service.php [hunk 5]
```diff
@@ -1015,34 +1015,36 @@ private function mark_payment_authorized( $order, $intent_data ) {
...
-	 * @param WC_Order $order         Order object.
-	 * @param array    $intent_data   The intent data associated with this order.
+	 * @param WC_Order                          $order         Order object.
+	 * @param WC_Payments_API_Payment_Intention $intent        The intent instance.
 	 *
 	 * @return void
 	 */
-	private function mark_payment_capture_completed( $order, $intent_data ) {
-		$note = $this->generate_capture_success_note( $order, $intent_data['intent_id'], $intent_data['charge_id'] );
+	private function mark_payment_capture_completed( $order, $intent ) {
```
Comment (kalessil) [inline includes/class-wc-payments-order-service.php:1023]: The changes in this method are great; I really like how the OOP approach here improves readability and is more friendly to code analysis tools. Thanks!

**Review by kalessil — COMMENTED** [review]
Did the next review round and leaving clarifying questions to answer before I can do final testing round.

File: includes/class-wc-payments-captured-event-note.php (hunk 2 above)
Comment (mgascam) [reply includes/class-wc-payments-captured-event-note.php]: > The fixed part of fees is stored as an integer type

True, but please note that in the server, the type of `$fixed_rate` of the [`Fee` class is float](https://github.com/Automattic/woocommerce-payments-server/blob/f7ea2fac88a22b4dfebb4678067c4a9d47f85a29/server/wp-content/rest-api-plugins/endpoints/wcpay/fees/class-fee.php#L28). My understanding is that when [calculating the effective fee in the server](https://github.com/Automattic/woocommerce-payments-server/blob/20aaa433c157383a8c8b9260941566bdb3334e5a/server/wp-content/rest-api-plugins/endpoints/wcpay/service/class-fee-service.php#L331),  there's a [conversion for fx transactions](exchange_fee_currency) and the [fixed rate is returned as a float](https://github.com/Automattic/woocommerce-payments-server/blob/14c555fa3986fd3a88a5070b1b3332dab20dcde8/server/wp-content/rest-api-plugins/endpoints/wcpay/service/class-foreign-exchange-service.php#L75). Maybe we should address the rounding in the server side, I'm happy to raise a separate ticket to tackle it or keep working on it as part of https://github.com/Automattic/woocommerce-payments-server/pull/4620. Please let me know what you think.

File: client/payment-details/timeline/map-events.js (hunk 4 above)
Comment (mgascam) [reply client/payment-details/timeline/map-events.js]: Yes I tried to answer in previous comment, but in a nutshell it comes from the server where fixed_rate is defined as float

File: includes/class-wc-payments-order-service.php (hunk 5 above)
Comment (mgascam) [reply includes/class-wc-payments-order-service.php:1023]: Agree, thanks

File: includes/class-wc-payments-captured-event-note.php (hunk 3 above)
Comment (mgascam) [reply includes/class-wc-payments-captured-event-note.php]: Addressed here https://github.com/Automattic/woocommerce-payments/pull/8107/commits/bfb187892037f4f33f5999d8fde706419d895fae thanks!

**Comment by kalessil** [comment]
To recap on the path forward here (from Teams chat):
- in this PR, we won't address rounding issues (we should revert rounding)
- server-side should address the rounding issues, but in a separate server PR

@mgascam: please update this PR, and I'll do a final review and testing round this week.

File: includes/class-wc-payments-captured-event-note.php (hunk 2 above)
Comment (mgascam) [reply includes/class-wc-payments-captured-event-note.php]: @kalessil I removed the attempts to address rounding issues in https://github.com/Automattic/woocommerce-payments/pull/8107/commits/1611c26316bdbe4518397329f082ba80abe10cc4 , based on the  https://github.com/Automattic/woocommerce-payments/pull/8107#issuecomment-1931847759

File: includes/class-wc-payments-order-service.php [hunk 6]
```diff
@@ -1015,34 +1015,36 @@ private function mark_payment_authorized( $order, $intent_data ) {
...
-		$this->enqueue_add_fee_breakdown_to_order_notes( $order, $intent_data['intent_id'] );
+		$this->enqueue_add_fee_breakdown_to_order_notes( $order, $intent_id );
 
 		/**
 		 * If we have a status for the fraud outcome, we want to add the proper meta data.
 		 * If auth/capture is enabled and the transaction is allowed, it will be 'allow'.
 		 * If it was held for review for any reason, it will be 'review'.
 		 */
-		if ( '' !== $intent_data['fraud_outcome'] && Rule::is_valid_fraud_outcome_status( $intent_data['fraud_outcome'] ) ) {
+		if ( '' !== $fraud_outcome && Rule::is_valid_fraud_outcome_status( $fraud_outcome ) ) {
```
Comment (kalessil) [inline includes/class-wc-payments-order-service.php:1040]: Nit: we can place `$fraud_outcome = $intent->get_metadata()['fraud_outcome'] ?? '';` before this if-stement (better context bounding).

**Review by kalessil — APPROVED** [review]
LGTM and tests well!

@mgascam: please file a new server issue for rounding issues discovered here and start working on it when ready.

File: includes/class-wc-payments-order-service.php (hunk 6 above)
Comment (mgascam) [reply includes/class-wc-payments-order-service.php:1040]: Nice touch, done here https://github.com/Automattic/woocommerce-payments/pull/8107/commits/28d7730547aa740bdb808f17dac8fecf9457fda1 thanks

**Pull request merged**